### PR TITLE
Use pre-commit in CI

### LINKED ISSUE #4793: Further improvements around `pre-commit`
In the context of adding `pre-commit`, there are a few further improvements along those lines (assuming that #4792 gets approved):

- [x] We could and should run `pre-commit` in CI. This is relatively simple within `Github` (see https://pre-commit.ci). Running in Gitlab would have the advantage of short-circuiting the CI but might be more setup and less features (that we might or might not actually want).
- [x] Running this in CI would make more scripts from the `test/` folder obsolete (and could integrate the rest).
- [ ] `pre-commit` provides the option to [automatically enable `pre-commit` on repos](https://pre-commit.com/#automatically-enabling-pre-commit-on-repositories). Once the usage of `pre-commit` is established among the developers, this could be a further simplification of the workflow and a gentle nudge for the last non-converted. Is still not strictly required; it just changes the default.
- [x] `flake8` includes `pyflakes`, so we could simplify the CI (doing this might be obsolete if we go along with the first item).
- [x] In my previous projects, I found [`ruff`](https://astral.sh/ruff) to be quite a convenient replacement for linting and formatting tools in the Python space. It could be a simplification of CI and pre-commit hooks to switch to this.
- [x] There is [a plethora](https://pre-commit.com/hooks.html) of further tools that could be enabled with less than handful of lines and there are more that are not listed on the linked site. I've found a lot of useful tools there but on a large existing codebase as this it might be a non-negligible amount of work to enforce linter rules that were not followed before.

Feel free to discuss if and, if so, which of these features we want on what timescale.

### PR DESCRIPTION
**Pull request by chillenzer**
Implements most of the stuff suggested in #4793 and thereby closes #4793.

One very important hook I have left out: https://github.com/koalaman/shellcheck
This has a whole lot of great heuristics to write somewhat less fragile bash. But given the current discrepancy between its ideas of robust `bash` and the actual state of the repository it would need some automation to harmonise the two.

Also `ruff` has a lot to offer in terms of linting rules for Python. I haven't done any tinkering in that direction but there is a lot great advice in there.

### CODE REVIEW

**Comment by PrometheusPi** [comment]
@chillenzer there are still some CI issues.

**Comment by chillenzer** [comment]
I'm aware. It's a minor change in a Python script where I change `type(...) == ...` to `isinstance(...)` while following a linter's suggestion. Apparently, I forgot a `not` or something. Gotta check in detail.

**Comment by chillenzer** [comment]
The refactoring issue is fixed (more precisely: reverted). One further note: If files are touched as a whole, this most likely comes down to the fact that I fixed permissions of some files.

**Comment by chillenzer** [comment]
Hi, CI failed due to [connection issues](https://gitlab.com/hzdr/crp/picongpu/-/jobs/6214387641) (?) or the like. Please restart CI.

**Comment by PrometheusPi** [comment]
@chillenzer You changed /added 86 files, some seem to have no connection to the pre-commit for the CI, like this [atomic physics image](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4817/files#diff-ec588138b77b1e5a7c61e188a3748ab292d8d0eb17e7ff4f7a17824eb6c6d6fe)
Please have a look at your pull request and clean up the extra files

**Comment by psychocoderHPC** [comment]
> @chillenzer You changed /added 86 files, some seem to have no connection to the pre-commit for the CI, like this [atomic physics image](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4817/files#diff-ec588138b77b1e5a7c61e188a3748ab292d8d0eb17e7ff4f7a17824eb6c6d6fe) Please have a look at your pull request and clean up the extra files

Most likely he fixed the access rights of the file. @chillenzer Is this correct?

**Comment by chillenzer** [comment]
Yes, as already stated here:

> The refactoring issue is fixed (more precisely: reverted). One further note: If files are touched as a whole, this most likely comes down to the fact that I fixed permissions of some files.

File: lib/python/picongpu/extra/plugins/jupyter_widgets/base_widget.py [hunk 1]
```diff
@@ -359,8 +359,9 @@ def _update_available_sim_times(self):
             self.sim_time_slider.value = last_selected_val
         else:
             print(
-                "last selected time {0} not present with newly"
-                " selected sim_times, set to {1}".format(last_selected_val, common_sim_times[0])
+                "last selected time {0} not present with newly" " selected sim_times, set to {1}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/plugins/jupyter_widgets/base_widget.py]: This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

File: lib/python/picongpu/extra/plugins/data/transitionradiation.py [hunk 2]
```diff
@@ -257,8 +258,9 @@ def get_data(self, iteration=None, **kwargs):
                 theta = int(np.floor(maxIndex / len(self.phis)))
 
             print(
-                "Angular intensity distribution is sliced at theta={:.2e} "
-                "with omega={:.2e}.".format(self.thetas[theta], self.omegas[omega])
+                "Angular intensity distribution is sliced at theta={:.2e} " "with omega={:.2e}.".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/plugins/data/transitionradiation.py]: This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

File: lib/python/picongpu/extra/plugins/data/transitionradiation.py [hunk 3]
```diff
@@ -239,8 +239,9 @@ def get_data(self, iteration=None, **kwargs):
                 omega = 0
 
             print(
-                "Angular intensity distribution is sliced at phi={:.2e} "
-                "with omega={:.2e}.".format(self.phis[phi], self.omegas[omega])
+                "Angular intensity distribution is sliced at phi={:.2e} " "with omega={:.2e}.".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/plugins/data/transitionradiation.py]: This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

File: lib/python/picongpu/extra/plugins/data/png.py [hunk 4]
```diff
@@ -192,9 +192,9 @@ def _get_for_iteration(self, iteration, species, species_filter="all", axis=None
             # verify requested iterations exist
             if not set(iteration).issubset(available_iterations):
                 raise IndexError(
-                    "Iteration {} is not available!\n"
-                    "List of available iterations: \n"
-                    "{}".format(iteration, available_iterations)
+                    "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/plugins/data/png.py]: This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

File: lib/python/picongpu/extra/plugins/data/phase_space.py [hunk 5]
```diff
@@ -198,9 +198,9 @@ def _get_for_iteration(self, iteration, ps, species, species_filter="all", file_
             # verify requested iterations exist
             if not set(iteration).issubset(available_iterations):
                 raise IndexError(
-                    "Iteration {} is not available!\n"
-                    "List of available iterations: \n"
-                    "{}".format(iteration, available_iterations)
+                    "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/plugins/data/phase_space.py]: This line appears to be too long according to [PEP8](https://peps.python.org/pep-0008/#maximum-line-length), that's why there used to be a line break. 
How can this pas the CI?

File: lib/python/picongpu/extra/plugins/data/energy_histogram.py [hunk 6]
```diff
@@ -159,9 +159,9 @@ def _get_for_iteration(self, iteration, species, species_filter="all", include_o
         # verify requested iterations exist
         if not set(iteration).issubset(data.index.values):
             raise IndexError(
-                "Iteration {} is not available!\n"
-                "List of available iterations: \n"
-                "{}".format(iteration, data.index.values)
+                "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/plugins/data/energy_histogram.py]: here too

File: lib/python/picongpu/extra/plugins/data/emittance.py [hunk 7]
```diff
@@ -163,9 +163,9 @@ def _get_for_iteration(self, iteration, species, species_filter="all", **kwargs)
         # verify requested iterations exist
         if not set(iteration).issubset(data.index.values):
             raise IndexError(
-                "Iteration {} is not available!\n"
-                "List of available iterations: \n"
-                "{}".format(iteration, data.index.values)
+                "Iteration {} is not available!\n" "List of available iterations: \n" "{}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/plugins/data/emittance.py:166]: here too

File: lib/python/picongpu/extra/input/parameters.py [hunk 8]
```diff
@@ -145,16 +146,18 @@ def _check_input(self, vals):
...
+                    )
                 )
         else:
             # check for valid range
             res = all([self.pic_range[0] <= v <= self.pic_range[1] for v in vals])
             if not res:
                 raise ValueError(
-                    "Invalid values found! Values should be "
-                    "contained in {0} but are {1}!".format(self.pic_range, vals)
+                    "Invalid values found! Values should be " "contained in {0} but are {1}!".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/input/parameters.py]: here too

File: lib/python/picongpu/extra/input/parameters.py [hunk 9]
```diff
@@ -145,16 +146,18 @@ def _check_input(self, vals):
             res = all([v in self.pic_values for v in vals])
             if not res:
                 raise ValueError(
-                    "Invalid values found! Values should be elements of "
-                    "{0} but are {1}!".format(self.pic_values, vals)
+                    "Invalid values found! Values should be elements of " "{0} but are {1}!".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/extra/input/parameters.py]: here too

File: include/picongpu/plugins/openPMD/GetComponentsType.hpp [hunk 10]
```diff
@@ -1,50 +1,50 @@
...
-         */
-        template<>
-        struct GetComponentsType<bool>
-        {
-            using type = std::enable_if_t<sizeof(bool) == sizeof(char), char>;
-        };
-
-    } // namespace traits
-} // namespace pmacc
+/* Copyright 2021-2023 Sergei Bastrakov
```
Comment (PrometheusPi) [inline include/picongpu/plugins/openPMD/GetComponentsType.hpp:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

File: share/picongpu/examples/LaserWakefield/include/picongpu/param/incidentField.param [hunk 11]
```diff
@@ -1,342 +1,342 @@
...
-            constexpr int32_t POSITION[3][2] = {
-                {16, -16}, // x direction [negative, positive]
-                {16, -16}, // y direction [negative, positive]
-                {16, -16} // z direction [negative, positive]
-            };
-
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov, Rene Widera
```
Comment (PrometheusPi) [inline share/picongpu/examples/LaserWakefield/include/picongpu/param/incidentField.param:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

File: share/picongpu/examples/FoilLCT/include/picongpu/param/incidentField.param [hunk 12]
```diff
@@ -1,331 +1,331 @@
...
-            constexpr int32_t POSITION[3][2] = {
-                {0, -1}, // x direction [negative, positive]
-                {16, -16}, // y direction [negative, positive]
-                {0, -1} // z direction [negative, positive]
-            };
-
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov, Rene Widera
```
Comment (PrometheusPi) [inline share/picongpu/examples/FoilLCT/include/picongpu/param/incidentField.param:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

File: share/picongpu/examples/Bunch/include/picongpu/param/incidentField.param [hunk 13]
```diff
@@ -1,229 +1,229 @@
...
-            constexpr int32_t POSITION[3][2] = {
-                {0, -1}, // x direction [negative, positive]
-                {16, -16}, // y direction [negative, positive]
-                {0, -1} // z direction [negative, positive]
-            };
-
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov, Rene Widera
```
Comment (PrometheusPi) [inline share/picongpu/examples/Bunch/include/picongpu/param/incidentField.param:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

File: share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param [hunk 14]
```diff
@@ -1,603 +1,603 @@
...
-            constexpr int32_t POSITION[3][2] = {
-                {1, -1}, // x direction [negative, positive]
-                {1, -1}, // y direction [negative, positive]
-                {1, -1} // z direction [negative, positive]
-            };
-
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2014-2023 Axel Huebl, Alexander Debus, Richard Pausch, Sergei Bastrakov
```
Comment (PrometheusPi) [inline share/picongpu/benchmarks/TWEAC-FOM/include/picongpu/param/incidentField.param:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right. 
I will ignore these files.

File: lib/python/picongpu/pypicongpu/species/operation/operation.py [hunk 15]
```diff
@@ -180,8 +180,9 @@ def bake_species_attributes(self) -> None:
             duplicate_names = [name for name in attr_names if attr_names.count(name) > 1]
             if 0 != len(duplicate_names):
                 raise ValueError(
-                    "only on attribute per type is allowed per "
-                    "species, offending: {}".format(", ".join(duplicate_names))
+                    "only on attribute per type is allowed per " "species, offending: {}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/pypicongpu/species/operation/operation.py]: same here

File: lib/python/picongpu/pypicongpu/species/operation/momentum/drift.py [hunk 16]
```diff
@@ -45,8 +45,9 @@ def __check_vector_real(self, vector: typing.Tuple[float, float, float]) -> None
         for invalid in [math.inf, -math.inf, math.nan]:
             if invalid in vector:
                 raise ValueError(
-                    "vector may only contain real components, "
-                    "offending axis: {}".format(["x", "y", "z"][vector.index(invalid)])
+                    "vector may only contain real components, " "offending axis: {}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/pypicongpu/species/operation/momentum/drift.py]: same here

File: lib/python/picongpu/pypicongpu/species/initmanager.py [hunk 17]
```diff
@@ -237,8 +238,7 @@ def __check_species_dependencies_circular(self) -> None:
             # check: self in dependency closure?
             if species in dependency_closure:
                 raise RecursionError(
-                    "species {} is in circular dependency, "
-                    "all dependencies are: {}".format(
+                    "species {} is in circular dependency, " "all dependencies are: {}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/pypicongpu/species/initmanager.py]: same here

File: lib/python/picongpu/pypicongpu/species/initmanager.py [hunk 18]
```diff
@@ -336,10 +336,10 @@ def __check_constant_attribute_dependencies(self) -> None:
                         )
 
                     # actual check:
-                    assert (
-                        required_attr.PICONGPU_NAME in species_attr_names
-                    ), "constant {} of species {} requires attribute {} to " "be present, but it is not".format(
-                        constant, species.name, required_attr
+                    assert required_attr.PICONGPU_NAME in species_attr_names, (
+                        "constant {} of species {} requires attribute {} to " "be present, but it is not".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/pypicongpu/species/initmanager.py]: same here

File: lib/python/picongpu/pypicongpu/species/initmanager.py [hunk 19]
```diff
@@ -372,10 +372,10 @@ def __check_constant_constant_dependencies(self):
                         raise ReferenceError("constants may not depend on themselves")
 
                     # check if constant exists
-                    assert species.has_constant_of_type(
-                        required_constant
-                    ), "species {}: required constant {} not found, " "(required by constant {})".format(
-                        species.name, required_constant, constant
+                    assert species.has_constant_of_type(required_constant), (
+                        "species {}: required constant {} not found, " "(required by constant {})".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/pypicongpu/species/initmanager.py]: same here

File: lib/python/picongpu/picmi/simulation.py [hunk 20]
```diff
@@ -497,8 +496,9 @@ def write_input_file(self, file_name: str) -> None:
     def step(self, nsteps: int = 1):
         if nsteps != self.max_steps:
             raise ValueError(
-                "PIConGPU does not support stepwise running. Invoke step() "
-                "with max_steps (={})".format(self.max_steps)
+                "PIConGPU does not support stepwise running. Invoke step() " "with max_steps (={})".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/picmi/simulation.py]: same here

File: lib/python/picongpu/picmi/simulation.py [hunk 21]
```diff
@@ -375,8 +373,9 @@ def __get_electron_species(self) -> PicongpuPicmiSpecies:
                 electrons_name += "_"
 
             logging.info(
-                "no electron species for ionization available, "
-                "creating electrons with name: {}".format(electrons_name)
+                "no electron species for ionization available, " "creating electrons with name: {}".format(
```
Comment (PrometheusPi) [inline lib/python/picongpu/picmi/simulation.py]: same here

File: lib/python/picongpu/picmi/simulation.py [hunk 22]
```diff
@@ -99,7 +97,7 @@ def __init__(
         self,
         picongpu_template_dir: typing.Optional[typing.Union[str, pathlib.Path]] = None,
         picongpu_typical_ppc: typing.Optional[int] = None,
-        **kw
+        **kw,
```
Comment (PrometheusPi) [inline lib/python/picongpu/picmi/simulation.py:100]: not essential: why is a comma after the last argument preferred?

File: lib/python/picongpu/picmi/simulation.py [hunk 23]
```diff
@@ -70,9 +70,7 @@ def __yee_compute_cfl_or_delta_t(self) -> None:
 
         if self.time_step_size is not None and self.solver.cfl is not None:
             # both cfl & delta_t given -> check their compatibility
-            delta_t_from_cfl = self.solver.cfl / (
-                constants.c * sqrt(1 / delta_x**2 + 1 / delta_y**2 + 1 / delta_z**2)
-            )
+            delta_t_from_cfl = self.solver.cfl / (constants.c * sqrt(1 / delta_x**2 + 1 / delta_y**2 + 1 / delta_z**2))
```
Comment (PrometheusPi) [inline lib/python/picongpu/picmi/simulation.py:73]: same here

File: lib/python/picongpu/picmi/gaussian_laser.py [hunk 24]
```diff
@@ -58,7 +56,7 @@ def __init__(
             [16, -16],
             [16, -16],
         ],
-        **kw
+        **kw,
```
Comment (PrometheusPi) [inline lib/python/picongpu/picmi/gaussian_laser.py:59]: strange comma

File: lib/python/picongpu/picmi/gaussian_laser.py [hunk 25]
```diff
@@ -86,7 +84,7 @@ def __init__(
             polarization_direction,
             focal_position,
             centroid_position,
-            **kw
+            **kw,
```
Comment (PrometheusPi) [inline lib/python/picongpu/picmi/gaussian_laser.py:87]: another strange comma

File: include/picongpu/param/incidentField.param [hunk 26]
```diff
@@ -1,180 +1,180 @@
...
-            constexpr int32_t POSITION[3][2] = {
-                {16, -16}, // x direction [negative, positive]
-                {16, -16}, // y direction [negative, positive]
-                {16, -16} // z direction [negative, positive]
-            };
-
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2020-2023 Sergei Bastrakov
```
Comment (PrometheusPi) [inline include/picongpu/param/incidentField.param:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: include/picongpu/fields/incidentField/profiles/profiles.hpp [hunk 27]
```diff
@@ -1,29 +1,29 @@
...
-
-#include "picongpu/fields/incidentField/profiles/DispersivePulse.hpp"
-#include "picongpu/fields/incidentField/profiles/ExpRampWithPrepulse.hpp"
-#include "picongpu/fields/incidentField/profiles/Free.hpp"
-#include "picongpu/fields/incidentField/profiles/GaussianPulse.hpp"
-#include "picongpu/fields/incidentField/profiles/None.hpp"
-#include "picongpu/fields/incidentField/profiles/PlaneWave.hpp"
-#include "picongpu/fields/incidentField/profiles/Polynom.hpp"
-#include "picongpu/fields/incidentField/profiles/Wavepacket.hpp"
+/* Copyright 2022-2023 Sergei Bastrakov
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/profiles.hpp:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: include/picongpu/fields/incidentField/profiles/profiles.def [hunk 28]
```diff
@@ -1,31 +1,31 @@
...
-#include "picongpu/fields/incidentField/profiles/ExpRampWithPrepulse.def"
-#include "picongpu/fields/incidentField/profiles/Free.def"
-#include "picongpu/fields/incidentField/profiles/GaussianPulse.def"
-#include "picongpu/fields/incidentField/profiles/None.def"
-#include "picongpu/fields/incidentField/profiles/PlaneWave.def"
-#include "picongpu/fields/incidentField/profiles/Polynom.def"
-#include "picongpu/fields/incidentField/profiles/PulseFrontTilt.def"
-#include "picongpu/fields/incidentField/profiles/Traits.hpp"
-#include "picongpu/fields/incidentField/profiles/Wavepacket.def"
+/* Copyright 2022-2023 Sergei Bastrakov
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/profiles.def:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: include/picongpu/fields/incidentField/profiles/Wavepacket.hpp [hunk 29]
```diff
@@ -1,185 +1,185 @@
...
-                struct GetFunctorIncidentB<profiles::Wavepacket<T_Params>>
-                {
-                    using type = detail::ApproximateIncidentB<
-                        typename GetFunctorIncidentE<profiles::Wavepacket<T_Params>>::type>;
-                };
-            } // namespace detail
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Axel Huebl, Heiko Burau, Rene Widera, Richard Pausch,
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/Wavepacket.hpp:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: include/picongpu/fields/incidentField/profiles/Wavepacket.def [hunk 30]
```diff
@@ -1,87 +1,87 @@
...
-                 *                  The focus position should be at or near the generation surface as the
-                 *                  laser equations are formulated for this case.
-                 */
-                template<typename T_Params = defaults::WavepacketParam>
-                struct Wavepacket;
-            } // namespace profiles
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Axel Huebl, Heiko Burau, Rene Widera, Richard Pausch,
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/Wavepacket.def:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: include/picongpu/fields/incidentField/profiles/Traits.hpp [hunk 31]
```diff
@@ -1,44 +1,44 @@
...
-    };
-
-
-    /** short hand notation for @see MakePeriodicTransversalHuygensSurfaceContiguous */
-    template<typename T_Profile>
-    constexpr bool makePeriodicTransversalHuygensSurfaceContiguous
-        = MakePeriodicTransversalHuygensSurfaceContiguous<T_Profile>::value;
-
-} // namespace picongpu::fields::incidentField::profiles
+/* Copyright 2023-2023 Rene Widera
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/Traits.hpp:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: include/picongpu/fields/incidentField/profiles/PulseFrontTilt.def [hunk 32]
```diff
@@ -1,69 +1,69 @@
...
-                 *                  see members of defaults::PulseFrontTiltParam
-                 *                  for required members
-                 */
-                template<typename T_Params = defaults::PulseFrontTiltParam>
-                using PulseFrontTilt = GaussianPulse<T_Params>;
-            } // namespace profiles
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Anton Helm, Heiko Burau, Rene Widera, Richard Pausch,
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/PulseFrontTilt.def:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: include/picongpu/fields/incidentField/profiles/Polynom.hpp [hunk 33]
```diff
@@ -1,170 +1,170 @@
...
-                struct GetFunctorIncidentB<profiles::Polynom<T_Params>>
-                {
-                    using type = detail::ApproximateIncidentB<
-                        typename GetFunctorIncidentE<profiles::Polynom<T_Params>>::type>;
-                };
-            } // namespace detail
-        } // namespace incidentField
-    } // namespace fields
-} // namespace picongpu
+/* Copyright 2013-2023 Heiko Burau, Rene Widera, Richard Pausch, Axel Huebl, Sergei Bastrakov
```
Comment (PrometheusPi) [inline include/picongpu/fields/incidentField/profiles/Polynom.hpp:1]: This is one of those files where GitHub shows med that everything changes, but it does not show me a change in access right.
I will ignore these files.

File: .pre-commit-config.yaml [hunk 34]
```diff
@@ -7,23 +8,64 @@ repos:
...
+    - id: ruff
+      types_or: [ python, pyi, jupyter ]
+      # Rule E721 is about using isinstance()
+      # but lib/python/picongpu/pypicongpu/rendering/renderer.py
+      # is strict about types. Could be solved locally but I think this is fine.
+      args: [ --fix, --ignore, E721]
+    # Run the formatter.
+    - id: ruff-format
+      types_or: [ python, pyi, jupyter ]
+      args: ["--line-length", "120"]
```
Comment (PrometheusPi) [inline .pre-commit-config.yaml:71]: This is a change from our previous 79 characters per line (PEP8) limit

File: docs/COMMIT.md [hunk 35]
```diff
@@ -33,7 +33,9 @@ $ pre-commit install
 From now on, each commit in this clone of the repo will be checked. See [pre-commit](https://pre-commit.com/) for usage
 details. Some hints:
 
-- You can run all hooks on all files via `pre-commit run --all-files`.
+- You can run all hooks on all files via `pre-commit run --all-files [--hook-stage manual]`. The last argument
+  `--hook-stage manual` includes rather slow additional tests that are run by the CI but are considered too heavy-weight
+  to run before each commit.
 - If a check fails, oftentimes it will automatically apply the fixes (e.g. run `clang-format`). If there are no unstaged
```
Comment (PrometheusPi) [inline docs/COMMIT.md]: optional (not part of this PR): 
```suggestion
- If a check fails, the fixes are often applied automatically (e.g., run `clang-format`). If there are no unstaged
```

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
looks god to me (except that GitHub does not show right changes correctly all the times). 
Only the characters per line increase from 79 to 120 causes some wired code style that should be fixed (or we go back to 79 characters)

File: share/picongpu/examples/LaserWakefield/include/picongpu/param/incidentField.param (hunk 11 above)
Comment (chillenzer) [reply share/picongpu/examples/LaserWakefield/include/picongpu/param/incidentField.param:1]: `git blame` reveals that the change stems from f29c574a5 which removed CRLF line endings (as the commit message quite literally states). That's why each and every line is different. **They all have to end...** (That sounded way too ominous to _not _ end with it.)

File: lib/python/picongpu/picmi/gaussian_laser.py (hunk 24 above)
Comment (chillenzer) [reply lib/python/picongpu/picmi/gaussian_laser.py:59]: It's a convention used by `black` as well as `ruff format` (which this PR is attempting to switch to). The reasoning for this is concisely shown by example [here](https://stackoverflow.com/questions/12087742/should-i-add-a-trailing-comma-after-the-last-argument-in-a-function-call/17492103#17492103).

File: .pre-commit-config.yaml (hunk 34 above)
Comment (chillenzer) [reply .pre-commit-config.yaml:71]: Yes, this was agreed with @psychocoderHPC and I thought that it was already agreed upon elsewhere. But I've just tried to look it up but admittedly I could neither find it in `CONTRIBUTING.md` nor in our [Contributing repo](https://github.com/ComputationalRadiationPhysics/contributing). No strong feelings here. Actually, I would personally not change it.

**Review by PrometheusPi — APPROVED** [review]

**Review by psychocoderHPC — APPROVED** [review]
